File: practice/1-binary-search/binary_search.go
package main

import "fmt"

type Solution1 struct{}

// Solution1: 左闭右闭区间
func (s Solution1) search(nums []int, target int) int {
	left, right := 0, len(nums)-1

	for left <= right {
		middle := (left + right) / 2
		if nums[middle] < target {
			left = middle + 1
		} else if nums[middle] > target {
			right = middle - 1
		} else {
			return middle
		}
	}

	return -1
}

type Solution2 struct{}

// Solution2: 左闭右开区间
func (s Solution2) search(nums []int, target int) int {
	left, right := 0, len(nums)

	for left < right {
		middle := (left + right) / 2
		if nums[middle] < target {
			left = middle + 1
		} else if nums[middle] > target {
			right = middle
		} else {
			return middle
		}
	}

	return -1
}

func main() {
	s1 := Solution1{}
	sr := s1.search([]int{-1, 0, 3, 5, 9, 12}, 9)
	fmt.Println(sr)

	s2 := Solution2{}
	sc := s2.search([]int{-1, 0, 3, 5, 9, 12}, 9)
	fmt.Println(sc)
}

File: practice/1-binary-search/binary_search.py
from types import List


class Solution1:
    """解法一: 左闭右闭
    """

    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1

        while left <= right:
            middle = (left + right) // 2
            if nums[middle] < target:
                left = middle + 1
            elif nums[middle] > target:
                right = middle - 1
            else:
                return middle

        return -1


class Solution2:
    """解法二: 左闭右开
    """

    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums)

        while left < right:
            middle = left + (right - left) // 2
            if nums[middle] > target:
                right = middle
            elif nums[middle] < target:
                left = middle + 1
            else:
                return middle
            
        return -1
